inc/functions.php

```diff
@@ -566,47 +624,73 @@ function my_date($format, $stamp=0, $offset="", $ty=1, $adodb=false)
  */
 function my_mail($to, $subject, $message, $from="", $charset="", $headers="", $keep_alive=false, $format="text", $message_text="", $return_email="")
 {
-	global $mybb;
+	global $mybb, $plugins;
 	static $mail;
 
-	// Does our object not exist? Create it
-	if(!is_object($mail))
-	{
-		require_once MYBB_ROOT."inc/class_mailhandler.php";
+	// Get our mail handler.
+	$mail = &get_mybb_mailhandler();
 
-		if($mybb->settings['mail_handler'] == 'smtp')
-		{
-			require_once MYBB_ROOT."inc/mailhandlers/smtp.php";
-			$mail = new SmtpMail();
-		}
-		else
-		{
-			require_once MYBB_ROOT."inc/mailhandlers/php.php";
-			$mail = new PhpMail();
-		}
-	}
-
-	// Using SMTP based mail
-	if($mybb->settings['mail_handler'] == 'smtp')
+	// If MyBB's built-in SMTP mail handler is used, set the keep alive bit accordingly.
+	if(isset($mybb->settings['mail_handler']) && $mybb->settings['mail_handler'] == 'smtp' && $keep_alive == true && !empty($mail->keep_alive))
 	{
-		if($keep_alive == true)
+		require_once MYBB_ROOT . "inc/class_mailhandler.php";
+		require_once MYBB_ROOT . "inc/mailhandlers/smtp.php";
+		if($mail instanceof MailHandler && $mail instanceof SmtpMail)
 		{
 			$mail->keep_alive = true;
 		}
 	}
 
-	// Using PHP based mail()
-	else
+	$my_mail_parameters = compact("to", "subject", "message", "from", "charset", "headers", "keep_alive", "format", "message_text", "return_email");
+
+	$my_mail_parameters = $plugins->run_hooks('my_mail_pre_build_message', $my_mail_parameters);
+
+	if(is_array($my_mail_parameters))
 	{
-		if($mybb->settings['mail_parameters'] != '')
+		foreach($my_mail_parameters as $key => $value)
 		{
-			$mail->additional_parameters = $mybb->settings['mail_parameters'];
+			// Using extract() here may not be safe, so we work it out manually.
+			if(in_array($key, array("to", "subject", "message", "from", "charset", "headers", "keep_alive", "format", "message_text", "return_email")))
+			{
+				$$key = $value;
+			}
 		}
 	}
 
-	// Build and send
+	// Build the mail message.
 	$mail->build_message($to, $subject, $message, $from, $charset, $headers, $format, $message_text, $return_email);
-	return $mail->send();
+
+	// Following variables will help sequential plugins to determine how to process the hook.
+	// Mark this variable true if the hooked plugin has sent the mail, otherwise don't modify it.
+	$is_mail_sent = false;
+	// Mark this variable false if the hooked plugin doesn't suggest sequential plugins to continue processing.
+	$continue_process = true;
+
+	$my_mail_parameters = compact("to", "subject", "message", "from", "charset", "headers", "keep_alive", "format", "message_text", "return_email", "is_mail_sent", "continue_process");
```

Comment:
Why aren't we passing by reference here?

Reply:
Sorry I don't quite understand you here. Did you mean passing the variable to the following hook by reference?

Reply:
I believe this is the intention:

```suggestion
	$my_mail_parameters = array(
		'to' => &$to,
		'subject' => &$subject,
		'message' => &$message,
		'from' => &$from,
		'charset' => &$charset,
		'headers => &$headers,
		'keep_alive' => &$keep_alive,
		'format' => &$format,
		'message_text' => &$message_text,
		'return_email' => &$return_email,
		'is_mail_sent' => &$is_mail_sent,
		'continue_process' => &$continue_process,
	);
```

Reply:
The suggestion has been made, and the variable is much simpler and the code looks much cleaner now. Had a quick test with no plugin hooked and the pre send hook, all went well.